_drafts/ble-throughput-primer.md

```diff
@@ -0,0 +1,288 @@
+
+{:.no_toc}
+
+#### Maximum Supported ATT MTU Size
+
+As we discussed [above](#max-throughput-gatt), the ATT MTU size has an impact on the max throughput which can be realized when using GATT due to protocol overhead. Some chips only support the default size (especially older Android Phones). When first testing a chip it's important to check and make sure larger packets can be sent!
+
+{:.no_toc}
+
+#### Coexistence
+
+A lot of devices share the 2.4GHz antenna for several different RF protocols. For example, on a phone this would typically be Bluetooth Classic (used for things like audio streaming to speakers), Bluetooth LE & WiFi. Additionally, sometimes for Bluetooth there will be multiple simultaneous connections that need to be managed.
+
+When multiple devices are sharing the same radio, the software stack on the device will assign some prioritization between the different protocols. For example, the scheduler in a phones may prioritize traffic based on whether the app using the connection is in the foreground or background or streaming audio.
+
+{:.no_toc}
+
+#### Reliability
+
+Counterintuitively, even though the Link Layer of BLE is _reliable_, packet loss is still something to be concerned about for BLE. This is because many many stacks drop data within the software stack. For example, BLE messages get queued up in the stack and when the heap holding the packets runs out of memory, some stacks will silently drop data. This means if you are sending large amounts of data over BLE you will usually want to add some sort of reliability layer that can detect & retransmit packets when data is dropped. The way this is implemented can have sizeable impacts on throughput. For example, if every single packet sent requires an ACK before another packet is sent, you'll typically wind up adding a connection interval worth of latency getting the ACK out, effectively halving the max throughput which can be achieved.
+
+## Bluetooth Low Energy Specification Updates Impacting Throughput
+
+Over the years, the Bluetooth SIG has added a number of additions to the Low Energy Specification to help improve the throughput that can be achieved. In this section we explore these settings
+
+> CAUTION: Even though some of these features have been around for a while, xssupport within the ecosystem still varies. If you are planning to leverage these features, I'd strongly recommend having away to dynamically enable/disable them in your software stack. While the Bluetooth Specification has a robust set of hardware level BLE "Direct Test"s a device must pass to be certified, there really is no thorough test of the software stack itself. This in part has contributed to very buggy BLE software stacks getting shipped and numerous interoperability issues when different Bluetooth Chips try to talk to one another.
```

Comment:
typo: xssuport => support

Reply:
I really need to setup that appointment at the apple store to get a new keyboard :/